archive/oxford_invariants/2021summer/week3_p1.lean

```diff
@@ -0,0 +1,137 @@
+  `ih` will be the induction hypothesis -/
+  induction n with n ih,
+  /- Base case
+  Claim that the sum equals `1`-/
+  { refine ⟨1, _, _⟩,
+    -- Check that this indeed equals the sum
+    { rw [nat.cast_one, finset.sum_range_one, div_self],
+      exact (mul_pos (a_pos 0) (a_pos 1)).ne' },
+    -- Check the modulo condition
+    { rw [mul_one, nat.sub_self],
+      exact dvd_zero _ } },
+  /- Induction step
+  `b` is the value of the previous sum as a natural, `hb` is the proof that it is indeed the value,
+  and `han` is the divisibility condition -/
+  obtain ⟨b, hb, han⟩ := ih (λ i hi, ha i $ nat.lt_succ_of_lt hi),
+  specialize ha n n.lt_succ_self,
+  have ha₀ : a 0 ≤ a n * b, -- Needing this is an artifact of `ℕ`-substraction.
+  { rw [←@nat.cast_le ℚ, nat.cast_mul, hb, ←div_le_iff' (a_pos _),
+      ←mul_div_mul_right _ _ (a_pos $ n + 1).ne'],
+    suffices h : ∀ i, i ∈ finset.range (n + 1) → 0 ≤ (a 0 : ℚ) * a (n + 1) / (a i * a (i + 1)),
+    { exact finset.single_le_sum h (finset.self_mem_range_succ n) },
+    refine (λ i _, div_nonneg _ _); refine mul_nonneg _ _; exact nat.cast_nonneg _ },
+  -- Claim that the  `(aₙ + aₙ₊₂)/aₙ₊₁ * b - (aₙ * b - a₀)/aₙ₊₁`
+  refine ⟨(a n + a (n + 2))/ a (n + 1) * b - (a n * b - a 0) / a (n + 1), _, _⟩,
+  -- Check that this indeed equals the sum
+  { calc
+      (((a n + a (n + 2)) / a (n + 1) * b - (a n * b - a 0) / a (n + 1) : ℕ) : ℚ)
+        = (a n + a (n + 2)) / a (n + 1) * b - (a n * b - a 0) / a (n + 1) : begin
+          rw [nat.cast_sub, nat.cast_mul, nat.cast_dvd_char_zero ha, nat.cast_add,
+            nat.cast_dvd_char_zero han, nat.cast_sub ha₀, nat.cast_mul],
+            { apply_instance }, -- char_zero ℚ
+            { apply_instance }, -- char_zero ℚ
+            { refine nat.div_le_of_le_mul _,
+              rw [←mul_assoc, nat.mul_div_cancel' ha, add_mul],
+              exact (nat.sub_le_self _ _).trans (nat.le_add_right _ _) }
+        end
+    ... = a (n + 2) / a (n + 1) * b + (a 0 * a (n + 2)) / (a (n + 1) * a (n + 2))
+        : by rw [add_div, add_mul, sub_div, mul_div_right_comm, add_sub_sub_cancel,
+            mul_div_mul_right _ _ (a_pos _).ne']
+    ... = ∑ (i : ℕ) in finset.range (n + 2), a 0 * a (n + 2) / (a i * a (i + 1))
+        : begin
+          rw [finset.sum_range_succ, hb, finset.mul_sum],
+          congr, ext i,
+          rw [←mul_div_assoc, ←mul_div_right_comm, mul_div_assoc, mul_div_cancel _ (a_pos _).ne',
+            mul_comm],
+        end },
+  -- Check the modulo condition
+  { rw [nat.mul_sub_left_distrib, ← mul_assoc, nat.mul_div_cancel' ha, add_mul,
+      nat.mul_div_cancel' han, nat.add_sub_sub_cancel ha₀, nat.add_sub_cancel],
+    exact dvd_mul_right _ _ }
+end
+
+theorem week3_p1 {n : ℕ} (hn : 1 ≤ n) (a : ℕ → ℕ) (a_pos : ∀ i, 0 < a i)
+  (ha : ∀ i < n - 1, a (i + 1) ∣ a i + a (i + 2)) :
+  ∃ b : ℕ, (b : ℚ) = ∑ i in finset.range n, (a 0 * a n)/(a i * a (i + 1)) :=
+begin
+  cases n,
+  { exact (nat.not_succ_le_zero 0 hn).elim },
```

Comment:
I think you can turn this into only one line using `cases hn`, if you want to be extra stylish.

Reply:
Or `nat.exists_eq_succ_of_ne_zero`

Reply:
Welp, got rid of it altogether.